datasets/machine-learning-dataset-for-sdo/sdo_ml_indexer.py

```diff
@@ -54,32 +66,166 @@ def index(data_dir):
+        goes_path = goes_dir / Path('goes_ts.csv')
+        if not pathlib.Path.exists(goes_path):
+            try:
+                print(f"retrieving goes data for {goes_path}")
+                os.makedirs(goes_dir, exist_ok=True)
+                satellite = get_goes_satellite(t_start)
+                search_result = Fido.search(
+                    a.Time(t_start, t_end), a.Instrument.xrs, satellite)
+                download_result = Fido.fetch(search_result)
+                goes_ts = ts.TimeSeries(download_result)
+                if isinstance(goes_ts, list) and len(goes_ts) > 0:
+                    frames = []
+                    for goes_ts_frm in goes_ts:
+                        frames.append(goes_ts_frm.to_dataframe())
+                    goes_ts_df = pd.concat(frames)
+                else:
+                    goes_ts_df = goes_ts.to_dataframe()
+
+                goes_ts_df.index.name = 'timestamp'
+                goes_ts_df.to_csv(goes_path, index=True,
+                                  date_format=ISO_FORMAT)
+            except Exception as e:
+                print(e)
+                goes_ts_df = pd.DataFrame()
+        else:
+            try:
+                print(f"goes data for {goes_path} already present")
+                if goes_path in goes_cache:
+                    goes_ts_df = goes_cache[goes_path]
+                else:
+                    goes_ts_df = pd.read_csv(goes_path)
+                    goes_ts_df["timestamp"] = pd.to_datetime(
+                        goes_ts_df['timestamp'], format=ISO_FORMAT)
+                    goes_ts_df = goes_ts_df.set_index('timestamp', drop=False)
+                    goes_ts_df = goes_ts_df.sort_index()
+                    goes_cache[goes_path] = goes_ts_df
+
+            except Exception as e:
+                print(e)
+                goes_ts_df = pd.DataFrame()
+
+    return goes_ts_df
+
+
+def get_goes_satellite(date):
+    # https://www.ngdc.noaa.gov/stp/satellite/goes-r.html
+    # https://www.ngdc.noaa.gov/stp/satellite/goes/index.html
+
+    # Since March 2020, data prior to GOES 15 (incl) is no longer supported by NOAA and GOES 16 and 17 data is now provided.
+    # GOES 16 and 17 are part of the GOES-R series and provide XRS data at a better time resolution (1s).
+    # GOES 16 has been taking observations from 2017, and GOES 17 since 2018
+
+    if date >= dt.datetime(2018, 6, 1):
+        return a.goes.SatelliteNumber(17)
+
+    if date >= dt.datetime(2017, 2, 7):
+        return a.goes.SatelliteNumber(16)
+
+    if date <= dt.datetime(2010, 9, 1):
+        return a.goes.SatelliteNumber(14)
+
+    return a.goes.SatelliteNumber(15)
+
+
+def get_goes_at(at, goes_dir):
+    # https://github.com/sunpy/sunpy/blob/master/sunpy/timeseries/sources/goes.py
+    # https://ngdc.noaa.gov/stp/satellite/goes/doc/GOES_XRS_readme.pdf
+    goes_ts_df = fetch_goes_metadata(at, at, goes_dir)
+    goes_at_time = goes_ts_df.iloc[goes_ts_df.index.get_loc(
```

Comment:
the current algorithm takes the closest value, should we think about a window aggregation?